include/aikido/control/detail/KinematicSimulationJointCommandExecutor-impl.hpp

```diff
@@ -0,0 +1,211 @@
+#include <dart/common/Console.hpp>
+
+#include "aikido/common/memory.hpp"
+
+namespace aikido {
+namespace control {
+
+//==============================================================================
+extern template class KinematicSimulationJointCommandExecutor<
+    ExecutorType::POSITION>;
+
+extern template class KinematicSimulationJointCommandExecutor<
+    ExecutorType::VELOCITY>;
+
+//==============================================================================
+template <ExecutorType T>
+KinematicSimulationJointCommandExecutor<
+    T>::KinematicSimulationJointCommandExecutor(::dart::dynamics::SkeletonPtr
+                                                    skeleton)
+  : JointCommandExecutor<T>(skeletonToJointNames(skeleton))
+  , mSkeleton{std::move(skeleton)}
+  , mInProgress{false}
+  , mPromise{nullptr}
+  , mMutex{}
+{
+  if (!mSkeleton)
+    throw std::invalid_argument("Skeleton is null.");
+}
+
+//==============================================================================
+template <ExecutorType T>
+KinematicSimulationJointCommandExecutor<
+    T>::~KinematicSimulationJointCommandExecutor()
+{
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+    DART_UNUSED(lock); // Suppress unused variable warning
+
+    if (mInProgress)
+    {
+      mInProgress = false;
+      mPromise->set_exception(
+          std::make_exception_ptr(std::runtime_error("Command canceled.")));
+    }
+    this->stop();
+  }
+}
+
+//==============================================================================
+template <ExecutorType T>
+std::future<int> KinematicSimulationJointCommandExecutor<T>::execute(
+    const std::vector<double> command,
+    const std::chrono::duration<double>& timeout)
+{
+
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+    DART_UNUSED(lock); // Suppress unused variable warning
+    auto promise = std::promise<int>();
+
+    if (command.size() != this->mJoints.size())
+    {
+      promise.set_exception(std::make_exception_ptr(
+          std::runtime_error("DOF of command does not match DOF of joints.")));
+      return promise.get_future();
+    }
+
+    if (mInProgress)
+    {
+      // Overwrite previous command
+      mCommand.clear();
+      mPromise->set_exception(
+          std::make_exception_ptr(std::runtime_error("Command canceled.")));
+    }
+
+    mPromise.reset(new std::promise<int>());
+
+    mCommand = command;
+    mInProgress = true;
+    mExecutionStartTime = std::chrono::system_clock::now();
+    mStartPosition.clear();
+    mTimeout = timeout;
+    switch (T)
+    {
+      case ExecutorType::VELOCITY:
+
+        for (size_t i = 0; i < mSkeleton->getDofs().size(); i++)
+        {
+          auto dof = mSkeleton->getDof(i);
+          dof->setVelocity(command[i]);
```

Comment:
Is `setVelocity` necessary if we're manually setting the position in `step`?

Reply:
Not for the purpose of this class, but the client could be making some calculates based on the joint's velocity, so we want it to be accurately reflected here.